### PR TITLE
feat!: re-write to TypeScript

### PR DESCRIPTION
**Pull request by Tofix26**

### CODE REVIEW

**Comment by SrIzan10** [comment]
Did you run prettier?

**Review by Murtatrxx — APPROVED** [review]
🎊

**Review by SrIzan10 — APPROVED** [review]

**Comment by SrIzan10** [comment]
@EvolutionX-10 is going to have a good time reading this pr lol

**Comment by Tofix26** [comment]
> Did you run prettier?

Didn't before but now fixed it

File: src/commands/init.ts [hunk 1]
```diff
@@ -19,18 +19,19 @@ import { editDirs, editMain } from '../utilities/edits.js';
 import { writeFile } from 'fs/promises';
 const { prompt } = prompts;
 
-/**
- * @param {{ y: string; sync: string; }} flags
- */
-export async function init(flags) {
+interface flags {
```
Comment (jacoobes) [inline src/commands/init.ts:22]: types convention is PascalCase, flags -> Flags

**Comment by SrIzan10** [comment]
Ok nice

File: tsconfig.json [hunk 2]
```diff
@@ -1,17 +1,16 @@
 {
```
Comment (jacoobes) [inline tsconfig.json:1]: Should we generate d.ts files?
(setting declaration: true)

File: src/commands/init.ts [hunk 3]
```diff
@@ -97,11 +95,12 @@ export async function init(flags) {
 	process.exit(0);
 }
 
-/**
- * It initializes git
- * @param data - The data object that contains the name of the project.
- */
-async function git(data) {
+interface data {
```
Comment (jacoobes) [inline src/commands/init.ts:98]: same here, PascalCase convention

File: package.json [hunk 4]
```diff
@@ -2,9 +2,9 @@
 	"name": "@sern/cli",
 	"version": "0.4.2",
```
Comment (jacoobes) [inline package.json:3]: bump cur -> 0.5.2

**Review by jacoobes — CHANGES_REQUESTED** [review]

**Comment by Murtatrxx** [comment]
Consider using conventional commits.

**Review by Murtatrxx — DISMISSED** [review]

**Comment by Tofix26** [comment]
> Consider using conventional commits.

forgot again :p

**Review by Murtatrxx — DISMISSED** [review]

**Review by EvolutionX-10 — CHANGES_REQUESTED** [review]

File: package.json [hunk 5]
```diff
@@ -1,10 +1,10 @@
 {
 	"name": "@sern/cli",
-	"version": "0.4.2",
+	"version": "0.5",
```
[the remark this answers is not in the document]
Comment (Tofix26) [reply package.json]: done

File: src/commands/init.ts [hunk 6]
```diff
@@ -19,18 +19,19 @@ import { editDirs, editMain } from '../utilities/edits.js';
 import { writeFile } from 'fs/promises';
 const { prompt } = prompts;
 
-/**
- * @param {{ y: string; sync: string; }} flags
- */
-export async function init(flags) {
+interface Flags {
+	y: string;
+	sync: string;
```
Comment (EvolutionX-10) [inline src/commands/init.ts:24]:
```suggestion
	y: boolean;
	sync: boolean;
```

**Review by EvolutionX-10 — CHANGES_REQUESTED** [review]

**Comment by Tofix26** [comment]
fixed ^

File: src/commands/init.ts [hunk 7]
```diff
@@ -19,18 +19,19 @@ import { editDirs, editMain } from '../utilities/edits.js';
...
- */
-export async function init(flags) {
+interface Flags {
+	y: boolean;
+	sync: boolean;
+}
+
+export async function init(flags: Flags) {
 	let data;
 	let git_init;
```
Comment (EvolutionX-10) [inline src/commands/init.ts:29]: can you specify their types?

File: src/commands/init.ts [hunk 8]
```diff
@@ -19,18 +19,19 @@ import { editDirs, editMain } from '../utilities/edits.js';
...
-	let pm;
+	let pm = flags.sync ? undefined : flags.y ? 'npm' : await npm();
 
 	if (flags.y) {
 		const projectName = await prompt([name]);
```
Comment (EvolutionX-10) [inline src/commands/init.ts:33]: type!

File: src/commands/init.ts [hunk 9]
```diff
@@ -82,11 +80,11 @@ export async function init(flags) {
 	let choice;
```
Comment (EvolutionX-10) [inline src/commands/init.ts:80]: type!

File: src/commands/init.ts [hunk 10]
```diff
@@ -97,11 +95,12 @@ export async function init(flags) {
...
 }
 
-/**
- * It initializes git
- * @param data - The data object that contains the name of the project.
- */
-async function git(data) {
+interface Data {
+	name: string;
+}
```
Comment (EvolutionX-10) [inline src/commands/init.ts:100]: shift all the type definitions and interfaces at the bottom of the file

File: src/prompts/init.ts [hunk 11]
```diff
@@ -1,4 +1,5 @@
 import { blueBright } from 'colorette';
+import type { PromptObject } from 'prompts';
 
 export const lang = {
```
Comment (EvolutionX-10) [inline src/prompts/init.ts:4]: put the type PromptObject on every exported object

File: src/prompts/plugin.ts [hunk 12]
```diff
@@ -17,13 +17,17 @@ async function gimmechoices() {
...
-	return data.map(
-		(/** @type {{ name: string; download_url: string; }} */ e) => ({
-			title: e.name,
-			value: e.download_url,
-		})
-	);
+	interface data {
+		name: string;
+		download_url: string;
+	}
```
Comment (EvolutionX-10) [inline src/prompts/plugin.ts:23]: shift it to bottom of the file and rename it to `Data`

File: tsconfig.json [hunk 13]
```diff
@@ -1,17 +1,18 @@
...
 		"module": "ESNext",
-		"moduleResolution": "NodeNext",
-		"lib": ["es2017", "dom"],
-		"allowJs": true,
-		"checkJs": true,
-		"noEmit": true,
-		"strict": false,
-		"noImplicitThis": true,
-		"alwaysStrict": true,
+		"moduleResolution": "node",
```
Comment (EvolutionX-10) [inline tsconfig.json:5]:
```suggestion
		"moduleResolution": "Node",
```

**Review by EvolutionX-10 — CHANGES_REQUESTED** [review]
probably the final changes

File: src/commands/init.ts [hunk 14]
```diff
@@ -19,18 +19,19 @@ import { editDirs, editMain } from '../utilities/edits.js';
...
-	let data;
-	let git_init;
-	let pm;
+export async function init(flags: Flags) {
+	let data = {
+		name: 'project',
+		lang: 'typescript',
+		main_dir: 'src',
+		cmds_dir: 'commands',
+	} as PromptData;
```
Comment (EvolutionX-10) [inline src/commands/init.ts:28]: why not just this?
```suggestion
	let data: PromptData;
```

**Review by EvolutionX-10 — CHANGES_REQUESTED** [review]

File: src/prompts/plugin.ts [hunk 15]
```diff
@@ -1,7 +1,7 @@
 import { fetch } from 'undici';
 import { getLang } from '../utilities/getLang.js';
 
-function upperCase(string) {
+function upperCase(string: string) {
```
Comment (EvolutionX-10) [inline src/prompts/plugin.ts:4]:
```suggestion
function upperCase(string: string | null) {
```

And in 
https://github.com/Tofix26/sern-cli-ts-rewrite/blob/e0dd4af7f2ee609aa02f0ef07a461119bd3e6752/src/utilities/getLang.ts#L4-L8

Make it return type `Promise<string>` and you can remove the jsdoc type

**Review by EvolutionX-10 — CHANGES_REQUESTED** [review]

File: src/utilities/install.ts [hunk 16]
```diff
@@ -8,11 +8,13 @@ import ora from 'ora';
...
  * It installs dependencies from a package.json file
- * @param {'skip' | 'npm' | 'yarn'} choice - The package manager to use.
- * @param {string} name - The name of the project
- * @returns a promise.
+ * @param choice - The package manager to use.
+ * @param name - The name of the project
  */
-export async function installDeps(choice, name) {
+export async function installDeps(
+	choice: 'skip' | 'npm' | 'yarn',
```
Comment (jacoobes) [inline src/utilities/install.ts:15]: could we lift this into a new type called PackageManagerChoice = 'skip' | 'npm' | 'yarn',
this type is stated twice here and https://github.com/sern-handler/cli/pull/74/files#r963771486

**Review by jacoobes — CHANGES_REQUESTED** [review]

**Review by jacoobes — DISMISSED** [review]

**Review by EvolutionX-10 — APPROVED** [review]
For future me: Learn how to push effectively in other's branches for PRs

**Pull request merged**